File: Money/Money.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Money
{
    public class Money : Expression
    {
        public int amount;
        protected string currency;

        public Money(int amount, string currency)
        {
            this.amount = amount;
            this.currency = currency;
        }

        public string Currency
        {
            get { return currency; }
        }

        public Expression Times(int multiplier) {
            return new Money(amount * multiplier, currency);
        }

        public Expression Plus(Expression addend)
        {
            return new Sum(this, addend);
        }

        public Money Reduce(Bank bank, string to)
        {
            int rate = bank.Rate(currency, to);
            return new Money(amount / rate, to);
        }

        public override bool Equals(object obj)
        {
            Money money = (Money)obj;
            return amount == money.amount
                && Currency == money.Currency;
        }

        public override string ToString()
        {
            return amount + " " + currency;
        }

        public static Money MakeDollar(int amount)
        {
            return new Money(amount, "USD");
        }

        public static Money MakeFranc(int amount)
        {
            return new Money(amount, "CHF");
        }
    }
}

File: PyMoney/money.py
class Money:
	def __init__(self, amount, currency):
		self.amount = amount
		self.currency = currency
	
	def __eq__(self, other):
		return self.amount == other.amount \
			and self.currency == other.currency
	
	def __repr__(self):
		return "<%s %s>" % (self.amount, self.currency)
	
	@classmethod
	def dollar(cls, amount):
		return Money(amount, 'USD')
	
	@classmethod
	def franc(cls, amount):
		return Money(amount, 'CHF')
	
	def times(self, multiplier):
		return Money(self.amount * multiplier, self.currency)
	
	def plus(self, addend):
		return Sum(self, addend)
	
	def reduce(self, bank, to):
		return Money(self.amount / bank.rate(self.currency, to), to)

class Bank:
	def __init__(self):
		self.rates = {}

	def reduce(self, source, to):
		return source.reduce(self, to)
	
	def addRate(self, from_currency, to_currency, rate):
		self.rates[(from_currency, to_currency)] = rate
	
	def rate(self, from_currency, to_currency):
		if (from_currency == to_currency):
			return 1
		return self.rates[(from_currency, to_currency)]

class Sum:
	def __init__(self, augend, addend):
		self.augend = augend
		self.addend = addend

	def times(self, multiplier):
		return Sum(self.augend.times(multiplier), self.addend.times(multiplier))
	
	def plus(self, addend):
		return Sum(self, addend)
	
	def reduce(self, bank, to):
		amount = self.augend.reduce(bank, to).amount + self.addend.reduce(bank, to).amount
		return Money(amount, to)
